File: computer/parachute/core/claude_usage.py
"""
Claude usage tracking service.

Fetches usage limits from Claude's OAuth API using the OAuth token
from `claude setup-token` (CLAUDE_CODE_OAUTH_TOKEN).
"""

import logging
from dataclasses import dataclass
from datetime import datetime
from typing import Optional

import httpx

logger = logging.getLogger(__name__)

USAGE_API_URL = "https://api.anthropic.com/api/oauth/usage"


@dataclass
class UsageLimit:
    """A single usage limit bucket."""
    utilization: float  # Percentage used (0-100)
    resets_at: Optional[datetime]


@dataclass
class ExtraUsage:
    """Extra usage credits (for Max plan)."""
    is_enabled: bool
    monthly_limit: int
    used_credits: float
    utilization: float


@dataclass
class ClaudeUsage:
    """Complete Claude usage information."""
    five_hour: Optional[UsageLimit]
    seven_day: Optional[UsageLimit]
    seven_day_sonnet: Optional[UsageLimit]
    seven_day_opus: Optional[UsageLimit]
    extra_usage: Optional[ExtraUsage]
    subscription_type: Optional[str]  # "pro", "max", etc.
    rate_limit_tier: Optional[str]
    error: Optional[str] = None


def get_claude_token() -> Optional[str]:
    """
    Get the Claude OAuth token from settings.

    Returns the token string or None if not configured.
    """
    from parachute.config import get_settings
    return get_settings().claude_code_oauth_token


def _parse_datetime(s: Optional[str]) -> Optional[datetime]:
    """Parse ISO datetime string."""
    if not s:
        return None
    try:
        # Handle timezone-aware ISO format
        return datetime.fromisoformat(s.replace("Z", "+00:00"))
    except ValueError:
        return None


def _parse_limit(data: Optional[dict]) -> Optional[UsageLimit]:
    """Parse a usage limit from API response."""
    if not data:
        return None
    return UsageLimit(
        utilization=data.get("utilization", 0),
        resets_at=_parse_datetime(data.get("resets_at")),
    )


def _parse_extra_usage(data: Optional[dict]) -> Optional[ExtraUsage]:
    """Parse extra usage from API response."""
    if not data:
        return None
    # Only return ExtraUsage if it's actually enabled
    # When disabled, the other fields are null
    if not data.get("is_enabled", False):
        return None
    return ExtraUsage(
        is_enabled=True,
        monthly_limit=data.get("monthly_limit") or 0,
        used_credits=data.get("used_credits") or 0,
        utilization=data.get("utilization") or 0,
    )


async def fetch_claude_usage() -> ClaudeUsage:
    """
    Fetch current Claude usage from the API.

    Uses the OAuth token from settings (CLAUDE_CODE_OAUTH_TOKEN).

    Returns a ClaudeUsage object with current limits, or with an error
    message if the fetch failed.
    """
    access_token = get_claude_token()

    if not access_token:
        return ClaudeUsage(
            five_hour=None,
            seven_day=None,
            seven_day_sonnet=None,
            seven_day_opus=None,
            extra_usage=None,
            subscription_type=None,
            rate_limit_tier=None,
            error="CLAUDE_CODE_OAUTH_TOKEN not configured (run `claude setup-token`)",
        )

    # subscription_type and rate_limit_tier are not available from token alone
    subscription_type = None
    rate_limit_tier = None

    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(
                USAGE_API_URL,
                headers={
                    "Accept": "application/json",
                    "Content-Type": "application/json",
                    "User-Agent": "claude-code/2.0.31",
                    "Authorization": f"Bearer {access_token}",
                    "anthropic-beta": "oauth-2025-04-20",
                },
                timeout=10.0,
            )

            if response.status_code != 200:
                return ClaudeUsage(
                    five_hour=None,
                    seven_day=None,
                    seven_day_sonnet=None,
                    seven_day_opus=None,
                    extra_usage=None,
                    subscription_type=subscription_type,
                    rate_limit_tier=rate_limit_tier,
                    error=f"API request failed: {response.status_code}",
                )

            data = response.json()

            return ClaudeUsage(
                five_hour=_parse_limit(data.get("five_hour")),
                seven_day=_parse_limit(data.get("seven_day")),
                seven_day_sonnet=_parse_limit(data.get("seven_day_sonnet")),
                seven_day_opus=_parse_limit(data.get("seven_day_opus")),
                extra_usage=_parse_extra_usage(data.get("extra_usage")),
                subscription_type=subscription_type,
                rate_limit_tier=rate_limit_tier,
            )

    except httpx.TimeoutException:
        return ClaudeUsage(
            five_hour=None,
            seven_day=None,
            seven_day_sonnet=None,
            seven_day_opus=None,
            extra_usage=None,
            subscription_type=subscription_type,
            rate_limit_tier=rate_limit_tier,
            error="Request timed out",
        )
    except Exception as e:
        logger.exception("Error fetching Claude usage")
        return ClaudeUsage(
            five_hour=None,
            seven_day=None,
            seven_day_sonnet=None,
            seven_day_opus=None,
            extra_usage=None,
            subscription_type=subscription_type,
            rate_limit_tier=rate_limit_tier,
            error=str(e),
        )

File: app/lib/features/chat/models/claude_usage.dart
// Models for Claude usage tracking
//
// Represents usage limits from Claude's OAuth API.

/// A single usage limit bucket (5-hour, 7-day, etc.)
class UsageLimit {
  /// Percentage of the limit used (0-100)
  final double utilization;

  /// When this limit resets (ISO datetime string)
  final String? resetsAt;

  const UsageLimit({
    required this.utilization,
    this.resetsAt,
  });

  factory UsageLimit.fromJson(Map<String, dynamic> json) {
    return UsageLimit(
      utilization: (json['utilization'] as num?)?.toDouble() ?? 0,
      resetsAt: json['resets_at'] as String?,
    );
  }

  /// Parse the reset time to a DateTime
  DateTime? get resetsAtDateTime {
    if (resetsAt == null) return null;
    try {
      return DateTime.parse(resetsAt!);
    } catch (e) {
      return null;
    }
  }

  /// Get human-readable time until reset
  String get resetsIn {
    final resetTime = resetsAtDateTime;
    if (resetTime == null) return '';

    final now = DateTime.now().toUtc();
    final diff = resetTime.difference(now);

    if (diff.isNegative) return 'now';

    if (diff.inHours > 0) {
      return '${diff.inHours}h ${diff.inMinutes % 60}m';
    } else if (diff.inMinutes > 0) {
      return '${diff.inMinutes}m';
    } else {
      return '<1m';
    }
  }
}

class ExtraUsage {
  final bool isEnabled;
  final int monthlyLimit;
  final double usedCredits;
  final double utilization;

  const ExtraUsage({
    required this.isEnabled,
    required this.monthlyLimit,
    required this.usedCredits,
    required this.utilization,
  });

  factory ExtraUsage.fromJson(Map<String, dynamic> json) {
    return ExtraUsage(
      isEnabled: json['is_enabled'] as bool? ?? false,
      monthlyLimit: json['monthly_limit'] as int? ?? 0,
      usedCredits: (json['used_credits'] as num?)?.toDouble() ?? 0,
      utilization: (json['utilization'] as num?)?.toDouble() ?? 0,
    );
  }

  /// Remaining credits
  double get remaining => monthlyLimit - usedCredits;
}

class ClaudeUsage {
  /// 5-hour rolling window usage
  final UsageLimit? fiveHour;

  /// 7-day rolling window usage
  final UsageLimit? sevenDay;

  /// 7-day Sonnet-specific usage
  final UsageLimit? sevenDaySonnet;

  /// 7-day Opus-specific usage
  final UsageLimit? sevenDayOpus;

  /// Extra usage credits (Max plan)
  final ExtraUsage? extraUsage;

  /// Subscription type (e.g., "pro", "max")
  final String? subscriptionType;

  /// Rate limit tier
  final String? rateLimitTier;

  /// Error message if fetch failed
  final String? error;

  const ClaudeUsage({
    this.fiveHour,
    this.sevenDay,
    this.sevenDaySonnet,
    this.sevenDayOpus,
    this.extraUsage,
    this.subscriptionType,
    this.rateLimitTier,
    this.error,
  });

  factory ClaudeUsage.fromJson(Map<String, dynamic> json) {
    return ClaudeUsage(
      fiveHour: json['five_hour'] != null
          ? UsageLimit.fromJson(json['five_hour'] as Map<String, dynamic>)
          : null,
      sevenDay: json['seven_day'] != null
          ? UsageLimit.fromJson(json['seven_day'] as Map<String, dynamic>)
          : null,
      sevenDaySonnet: json['seven_day_sonnet'] != null
          ? UsageLimit.fromJson(json['seven_day_sonnet'] as Map<String, dynamic>)
          : null,
      sevenDayOpus: json['seven_day_opus'] != null
          ? UsageLimit.fromJson(json['seven_day_opus'] as Map<String, dynamic>)
          : null,
      extraUsage: json['extra_usage'] != null
          ? ExtraUsage.fromJson(json['extra_usage'] as Map<String, dynamic>)
          : null,
      subscriptionType: json['subscription_type'] as String?,
      rateLimitTier: json['rate_limit_tier'] as String?,
      error: json['error'] as String?,
    );
  }

  /// Whether this usage data contains an error
  bool get hasError => error != null && error!.isNotEmpty;

  /// Whether usage data is available
  bool get hasData => fiveHour != null || sevenDay != null;

  /// Friendly subscription name
  String get subscriptionName {
    switch (subscriptionType) {
      case 'max':
        return 'Max';
      case 'pro':
        return 'Pro';
      default:
        return subscriptionType ?? 'Unknown';
    }
  }
}
